open_spiel/python/games/liars_poker.py

```diff
@@ -0,0 +1,379 @@
+    self.deck = [_FULL_DECK[i] for i in range(self.num_digits)]
+
+  def new_initial_state(self):
+    """Returns a state corresponding to the start of a game."""
+    return LiarsPokerState(self)
+
+  def make_py_observer(self, iig_obs_type=None, params=None):
+    """Returns an object used for observing game state."""
+    return LiarsPokerObserver(
+      iig_obs_type or pyspiel.IIGObservationType(perfect_recall=False),
+      self.num_players(),
+      self.hand_length,
+      self.num_digits,
+      params)
+
+
+class LiarsPokerState(pyspiel.State):
+  """A python version of the Liars Poker state."""
+
+  def __init__(self, game):
+    """Constructor; should only be called by Game.new_initial_state."""
+    super().__init__(game)
+    # Game attributes
+    self._num_players = game.num_players()
+    self._hand_length = game.hand_length
+    self._num_digits = game.num_digits
+    self._deck = game.deck
+    self.hands = [[] for _ in range(self._num_players)]
+
+    # Action dynamics
+    total_possible_bets = game.hand_length * game.num_digits * self._num_players
+    self.bid_history = np.zeros((total_possible_bets, self._num_players))
+    self.challenge_history = np.zeros((total_possible_bets, self._num_players))
+    self._current_player = 0
+    self._bid_offset = len(Action)
+    self._max_bid = (self._hand_length * self._num_digits * self._num_players
+                     + self._bid_offset - 1)
+    self._bid_originator = -1 
+    self._current_bid = -1
+    self._num_challenges = 0
+    self.is_rebid = False
+
+    # Game over dynamics
+    self._game_over = False
+    self._winner = -1
+    self._loser = -1
+
+  def current_player(self):
+    """Returns id of the current player to act.
+    
+    The id is:
+      - TERMINAL if game is over.
+      - CHANCE if a player is drawing a number to fill out their hand.
+      - a number otherwise.
+    """
+    if self.is_terminal():
+      return pyspiel.PlayerId.TERMINAL
+    elif len(self.hands[self._num_players - 1]) < self._hand_length:
+      return pyspiel.PlayerId.CHANCE
+    else:
+      return self._current_player
+
+  def _is_challenge_possible(self):
+    """A challenge is possible once the first bid is made."""
+    return self._current_bid != -1
+
+  def _is_rebid_possible(self):
+    """A rebid is only possible when all players have challenged the original bid."""
+    return not self.is_rebid and self._num_challenges == self._num_players - 1
+
+  def _legal_actions(self, player):
+    """Returns a list of legal actions, sorted in ascending order."""
+    assert player >= 0
+    actions = []
+
+    if self._is_challenge_possible():
+      actions.append(Action.CHALLENGE)
+
+    if player != self._bid_originator or self._is_rebid_possible():
+      # Any move higher than the current bid is allowed.
+      # Bids start at 2 as 0 and 1 are for bid and challenge.
+      for b in range(max(self._bid_offset, self._current_bid + 1), self._max_bid + 1):
```

Comment:
I think this might be better readable by looping over the bid's without offset and finally e.g.
actions.append(bid + BID_ACTION_OFFSET)